# <center> Customer Churn Prediction using Artificial Neural Neworks </center>

### In this ANN project I am going to perform basic neural network operations starting from the very 1st step of dataset importing to final optimization. The following are the steps:
- Importing dataset
- Performing basic preprocessing operations such as Label Encoding and One Hot Encoding
- Standardizing the data using standard scaler
- creating the ANN model with appropriate parameters
- Training the ANN model on our dataset
- Evaluating the performance of the model with predicted values

### Part 1: Data Preprocessing

In [1]:
# importing relevant libraries
import numpy as np 
import pandas as pd 
import matplotlib.pyplot as plt 
import tensorflow as tf
from keras.models import Sequential
from keras.layers import Dense

In [2]:
# importing dataset 

df = pd.read_csv('churn_modelling.csv')

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 14 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   RowNumber        10000 non-null  int64  
 1   CustomerId       10000 non-null  int64  
 2   Surname          10000 non-null  object 
 3   CreditScore      10000 non-null  int64  
 4   Geography        10000 non-null  object 
 5   Gender           10000 non-null  object 
 6   Age              10000 non-null  int64  
 7   Tenure           10000 non-null  int64  
 8   Balance          10000 non-null  float64
 9   NumOfProducts    10000 non-null  int64  
 10  HasCrCard        10000 non-null  int64  
 11  IsActiveMember   10000 non-null  int64  
 12  EstimatedSalary  10000 non-null  float64
 13  Exited           10000 non-null  int64  
dtypes: float64(2), int64(9), object(3)
memory usage: 1.1+ MB


In [4]:
df.sample(2)

RowNumber  CustomerId  Surname  CreditScore Geography  Gender  Age  \
5631       5632    15782758  Ozerova          632    France    Male   40   
6251       6252    15614520    Smith          682    France  Female   37   

      Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
5631       5  147650.68              1          1               1   
6251       8  148580.12              1          1               0   

      EstimatedSalary  Exited  
5631        199674.83       0  
6251         35179.18       0

In [5]:
# dividing the dataset into dependent and independent variables by slicing. and while at it, removing those variables which has
# no impact on the decision of the model, like, RowNumber, CustomerID, Surname. 
x = df.iloc[:,3:-1].values
y = df.iloc[:,-1].values

In [6]:
print(x)

[[619 'France' 'Female' ... 1 1 101348.88]
 [608 'Spain' 'Female' ... 0 1 112542.58]
 [502 'France' 'Female' ... 1 0 113931.57]
 ...
 [709 'France' 'Female' ... 0 1 42085.58]
 [772 'Germany' 'Male' ... 1 0 92888.52]
 [792 'France' 'Female' ... 1 0 38190.78]]


#### Data Encoding

- we have categorical columns which are not ideal for a NN to train on, so we need to use dummy variables.
- we have 2 kinds of categorical variables here. Binary and Nominal. i.e., Gender is Binary - Male & Female
- geographical data which is the Country has no order in it so thats nominal variable.
- for Binary categorical variable we use Label Encoder
- for Nominal categorical variable we use One Hot Encoder

In [7]:
# Label Encoding:

from sklearn.preprocessing import LabelEncoder
le = LabelEncoder()

x[:,2] = le.fit_transform(x[:,2])

In [8]:
# One Hot Encoding:

# Column Transformer - is a function of sklearn.compose module which allows the transformation of different columns 
# separately and the features generated by the transformations are then concatenated into a single feature space

from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder

ct = ColumnTransformer(transformers = [('encoder',OneHotEncoder(),[1])],remainder = 'passthrough')
x = np.array(ct.fit_transform(x))

#### Performing data splitting and scaling of data

In [9]:
from sklearn.model_selection import train_test_split

x_train,x_test,y_train,y_test = train_test_split(x,y,test_size = 0.2, random_state = 0 )

In [10]:
# performing feature scaling

from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
x_train = scaler.fit_transform(x_train)
x_test = scaler.transform(x_test)

### Part 2: Building ANN Model

In [11]:
# initializing the model

ann = Sequential()

#### Building the input layer and 1st hidden layer

number of units = average of number of inputs and output <br>
in our case ; Units = (11+1)/2 = 6, <br>
this is basically the number of nodes in the hidden layer

In [12]:
ann.add(Dense(units=6,activation='relu'))

#### Building 2nd hidden layer

In [13]:
ann.add(Dense(units=6,activation='relu'))

#### Building output layer

In [14]:
ann.add(Dense(units=1,activation='sigmoid'))

### Part 3 : Training the ANN  

In [15]:
# model compiling :

In [16]:
ann.compile(optimizer = 'adam',loss = 'binary_crossentropy', metrics = ['accuracy'])

In [17]:
# model fitting:

In [18]:
with tf.device('/GPU:0'):
    ann.fit(x_train,y_train,batch_size=32, epochs=100)

Epoch 1/100
250/250 [==============================] - 2s 4ms/step - loss: 0.6578 - accuracy: 0.6275
Epoch 2/100
250/250 [==============================] - 1s 4ms/step - loss: 0.4864 - accuracy: 0.7969
Epoch 3/100
250/250 [==============================] - 1s 5ms/step - loss: 0.4533 - accuracy: 0.7993
Epoch 4/100
250/250 [==============================] - 1s 5ms/step - loss: 0.4371 - accuracy: 0.8049
Epoch 5/100
250/250 [==============================] - 1s 5ms/step - loss: 0.4286 - accuracy: 0.8091
Epoch 6/100
250/250 [==============================] - 1s 5ms/step - loss: 0.4239 - accuracy: 0.8105
Epoch 7/100
250/250 [==============================] - 1s 5ms/step - loss: 0.4203 - accuracy: 0.8135
Epoch 8/100
250/250 [==============================] - 1s 5ms/step - loss: 0.4171 - accuracy: 0.8154
Epoch 9/100
250/250 [==============================] - 1s 5ms/step - loss: 0.4131 - accuracy: 0.8175
Epoch 10/100
250/250 [==============================] - 1s 5ms/step - loss: 0.4084 - accura

### Part 4: Predicting Making the predictions and evaluating the model 😱

Now using this ANN model to predict if the customer with the following informations will leave the bank or not: 

Geography: France

Credit Score: 600

Gender: Male

Age: 40 years old

Tenure: 3 years

Balance: \$ 60000

Number of Products: 2

Does this customer have a credit card? Yes

Is this customer an Active Member: Yes

Estimated Salary: \$ 50000

So, should we say goodbye to that customer? 😰😰😰😰

In [19]:
# now that the model has been trained we need to predict new data using our model.
# passing all the values in an orderly fashion as per the dataset to predict churning 

print(ann.predict(scaler.transform([[1, 0, 0, 600, 1, 40, 3, 60000, 2, 1, 1, 50000]])) > 0.5)

1/1 [==============================] - 0s 109ms/step
[[False]]


Therefore, our ANN model predicts that this customer stays in the bank! 😌😌😌😌

**Important note:** Notice also that the "France" country was not input as a string in the last column but as "1, 0, 0" in the first three columns. That's because of course the predict method expects the one-hot-encoded values of the state, and as we see in the first row of the matrix of features X, "France" was encoded as "1, 0, 0". And be careful to include these values in the first three columns, because the dummy variables are always created in the first columns.

#### Making a Confusion matrix: 

In [20]:
y_pred = ann.predict(x_test)
y_pred = (y_pred > 0.5)
print(np.concatenate((y_pred.reshape(len(y_pred),1), y_test.reshape(len(y_test),1)),1))

63/63 [==============================] - 0s 1ms/step
[[0 0]
 [0 1]
 [0 0]
 ...
 [0 0]
 [0 0]
 [0 0]]


In [21]:
from sklearn.metrics import precision_score, \
    recall_score, confusion_matrix, \
    accuracy_score, f1_score


cm = confusion_matrix(y_test, y_pred)
print(cm)
print('Accuracy :' ,accuracy_score(y_test, y_pred))
print('Precision :' ,precision_score(y_test, y_pred, pos_label=0))
print('f1 score :' ,f1_score(y_test, y_pred, pos_label=0))
print('recall :' ,recall_score(y_test, y_pred, pos_label=0))

[[1514   81]
 [ 198  207]]
Accuracy : 0.8605
Precision : 0.8843457943925234
f1 score : 0.9156335046870275
recall : 0.9492163009404389


True Positives = 1514 <br>
False Positive = 81 <br>
True Negative = 198 <br>
False Negative = 207 <br>

## Conclusion, Business inferences and further scope from this project 

**Accuracy** of 86.05% signifies that the model can correctly predict churn or no churn for 86.05% of the cases. <br>
**Precision** of 88.43% implies that out of all the customers predicted to churn, 88.43% of them did churn, which is a crucial model reliability factor for data driven business decision making. <br>
**Spicificity** calculated as True Negative /(True Negative + False Positive) = 70.96% which basically means that the model correctly identifies customers who are not likely to churn.  <br>
**Recall** The model captures 94.92% of the actual churners correctly which is an important decision making factor for business. <br>
**f1 score** This score is denoting that how well is the model balancing **Precision** & **Recall**. And a score of 91.56 % signifies the model performs well in both correctly identifying churners and minimizing false alarms.<br>

### Business Value of the model: <br>**  

The customer churn prediction model, built using Artificial Neural Networks (ANNs), demonstrates a strong predictive capability with an **accuracy of 86.05%**, indicating reliable identification of customers at risk of leaving. Additionally:  

- **High Recall (94.92%)** ensures the model identifies the vast majority of customers likely to churn, allowing the business to proactively intervene and retain valuable customers.  
- **Precision (88.43%)** signifies that most of the identified churn-risk customers truly need attention, ensuring marketing or retention efforts are not wasted.  
- The **F1-score (91.56%)**, balancing precision and recall, highlights the model’s overall robustness in managing churn prediction effectively.  

By implementing this model:  
1. **Improve Customer Retention**: Identifying and addressing churn risks before customers leave helps reduce attrition rates, ensuring sustained revenue streams and long-term business growth.  
2. **Optimize Resource Allocation**: The high precision minimizes unnecessary expenditure by focusing retention strategies on genuinely at-risk customers.  
3. **Enhance Customer Experience**: Proactive engagement with at-risk customers can strengthen loyalty and foster stronger relationships.  
4. **Boost Profitability**: Retaining existing customers is more cost-effective than acquiring new ones, directly improving the bottom line.  

This model can act as a critical decision-support tool for retention teams, empowering them to make data-driven interventions that yield tangible ess outcomes.


- 

### Further scope of the project: <br>

1. Data Refinement:
    - Feature Engineering: Add more relevant features that may influence churn (e.g., customer interaction history, satisfaction scores, product usage patterns).
          Use domain knowledge to create meaningful derived metrics (e.g., lifetime value, inactivity duration).
    - Data Balancing: Address any class imbalance by techniques such as SMOTE (Synthetic Minority Oversampling Technique) or undersampling to ensure the model              doesn't favor the majority class.
    - Address Outliers and Missing Data: Clean the dataset further to handle anomalies that may skew predictions.
  
2. Model Optimization:
    - Hyperparameter Tuning: Optimize the ANN’s architecture, including the number of layers, neurons, learning rate, and activation functions using techniques like        grid search or Bayesian optimization.
    - Regularization: Apply L1/L2 regularization or dropout layers to prevent overfitting.
    - Advanced Architectures: Experiment with recurrent neural networks (RNNs) or attention mechanisms if temporal data (e.g., sequential transactions) is available.
  
3. Improve Training Data Quality:
    - Augment Data Sources: Integrate additional data from customer surveys, support tickets, or social media sentiments to capture latent factors influencing churn.
    - Behavioral Segmentation: Segment customers into cohorts based on demographics, usage, or profitability, and build tailored models for each group.